File: dsa-problems/LinkedLists/solutions/python/LNK-015-workshop-odd-even-grouping-stable.py
import sys

class ListNode:
    def __init__(self, val=0):
        self.val = val
        self.next = None

def group_odd_even_stable(head: ListNode) -> ListNode:
    odd_dummy = ListNode(0)
    even_dummy = ListNode(0)
    odd_tail = odd_dummy
    even_tail = even_dummy
    
    curr = head
    while curr:
        if curr.val % 2 != 0:
            odd_tail.next = curr
            odd_tail = odd_tail.next
        else:
            even_tail.next = curr
            even_tail = even_tail.next
        curr = curr.next
        
    even_tail.next = None
    odd_tail.next = even_dummy.next
    
    return odd_dummy.next

def main():
    input = sys.stdin.read
    data = input().split()
    if not data:
        return
    
    iterator = iter(data)
    try:
        n = int(next(iterator))
        dummy = ListNode()
        cur = dummy
        for _ in range(n):
            cur.next = ListNode(int(next(iterator)))
            cur = cur.next
            
        head = group_odd_even_stable(dummy.next)
        
        out = []
        while head:
            out.append(str(head.val))
            head = head.next
        print(" ".join(out))
    except StopIteration:
        pass

if __name__ == "__main__":
    main()

File: dsa-problems/LinkedLists/solutions/javascript/LNK-015-workshop-odd-even-grouping-stable.js
const readline = require("readline");

class ListNode {
  constructor(val) {
    this.val = val;
    this.next = null;
  }
}

function groupOddEvenStable(head) {
  const oddDummy = new ListNode(0);
  const evenDummy = new ListNode(0);
  let oddTail = oddDummy;
  let evenTail = evenDummy;

  let curr = head;
  while (curr) {
    if (curr.val % 2 !== 0) {
      oddTail.next = curr;
      oddTail = oddTail.next;
    } else {
      evenTail.next = curr;
      evenTail = evenTail.next;
    }
    curr = curr.next;
  }

  evenTail.next = null;
  oddTail.next = evenDummy.next;

  return oddDummy.next;
}

const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
});

let data = [];
rl.on("line", (line) => data.push(...line.trim().split(/\s+/).filter(x => x)));
rl.on("close", () => {
  if (data.length === 0) return;
  let idx = 0;
  const n = parseInt(data[idx++], 10);
  
  const dummy = new ListNode(0);
  let cur = dummy;
  for (let i = 0; i < n; i++) {
    cur.next = new ListNode(parseInt(data[idx++], 10));
    cur = cur.next;
  }

  let head = groupOddEvenStable(dummy.next);
  const out = [];
  while (head) {
    out.push(head.val);
    head = head.next;
  }
  console.log(out.join(" "));
});
